src/lib/tls/asio/asio_stream.h

```diff
@@ -214,31 +220,23 @@ class Stream
          {
          setup_native_handle(side, ec);
 
-         // send client hello, which was written to the send buffer on client instantiation
-         send_pending_encrypted_data(ec);
+         if(side == CLIENT)
```

Comment:
This check isn't actually needed, as `send_pending_encrypted_data` won't have anything to send if there's no client native handle that put the client-hello into the send buffer.
In fact, the async handshake doesn't explicitly check this. Should we try to `send_pending...` unconditionally?

Reply:
It's not really possible to commit to the send_buffer before doing the handshake, because the `handshake()` is what creates the underlying client or server `Channel`. However, I found that trying to do so was not handled gracefully: https://github.com/randombit/botan/pull/2229/commits/b8f2f06631885a7933ec700e42f0d094f1876a30

Is `Invalid_State` the right exception to use here?

Reply:
```
// Invalid state was encountered. A request was made on an object while the
// object was in a state where the operation cannot be performed.
```

Sounds good to me 👍